# Week 12
## Loading Data

In [64]:
import pandas as pd

In [65]:
shootings_df = pd.read_csv('../Data/Shootings_(2006-Present)_20260401.csv')              # loading three dataset files
offenders_df = pd.read_csv('../Data/Shooting_Offenders_(2006-Present)_20260401.csv')
victims_df = pd.read_csv('../Data/Shooting_Victims_(2006-Present)_20260401.csv')

In [67]:
combined_df = pd.merge(shootings_df, offenders_df, on='INCIDENT_KEY', how='outer')             # merging the datasets to a combined dataset
combined_df = pd.merge(combined_df, victims_df, on='INCIDENT_KEY', how='outer')

## Exploration

### Column Descriptions for Shootings Dataset

- `INCIDENT_KEY` - Randomly generated persistent ID for each incident

- `OCCUR_DATE` - Exact date of the shooting incident

- `OCCUR_TIME` - Exact time of the shooting incident

- `BORO` - Borough where the shooting incident occurred

- `LOC_OF_OCCUR_DESC` - Flag indicating where incident occurred

- `PRECINCT` - Precinct where the shooting incident occurred

- `JURISDICTION_CODE` - Jurisdiction where the shooting incident occurred. Jurisdiction codes 0(Patrol), 1(Transit) and 2(Housing) represent NYPD Codes 3 and above represent non NYPD jurisdictions

- `LOC_CLASSFCTN_DESC` - Category describing type of location

- `LOCATION_DESC` - Category for specific type of location of the shooting incident

- `X_COORD_CD` - Midblock X-coordinate for New York State Plane Coordinate System, Long Island Zone, NAD 83, units feet (FIPS 3104)

- `Y_COORD_CD` - Midblock Y-coordinate for New York State Plane Coordinate System, Long Island Zone, NAD 83, units feet (FIPS 3104)

- `Latitude` - Midblock Latitude coordinate for Global Coordinate System, WGS 1984, decimal degrees (EPSG 4326)

- `Longitude` - Midblock Longitude coordinate for Global Coordinate System, WGS 1984, decimal degrees (EPSG 4326)



### Column Descriptions for Offenders Dataset

- `INCIDENT_KEY` - Randomly generated persistent ID for each incident

- `PERP_ID` - Randomly generated persistent ID for each suspect (perpetrator)

- `PERP_AGE_GROUP` - Perpetrator’s age within a category

- `PERP_SEX` - Perpetrator’s sex description

- `PERP_RACE` - Perpetrator’s race description

### Column Descriptions for Victims Dataset

- `INCIDENT_KEY` - Randomly generated persistent ID for each incident

- `VICTIM_ID` - Randomly generated persistent ID for each victim

- `VICTIM_AGE_GROUP` - Victim’s age within a category

- `VICTIM_SEX` - Victim’s sex description

- `VICTIM_RACE` - Victim’s race description

- `STAT_MURDER_FLG` - Shooting resulted in the victim’s death which would be counted as a murder

In [68]:
combined_df.head(10)

,INCIDENT_KEY,OCCUR_DATE,OCCUR_TIME,BORO,LOC_OF_OCCUR_DESC,PRECINCT,JURISDICTION_CODE,LOC_CLASSFCTN_DESC,LOCATION_DESC,X_COORD_CD,...,Longitude,PERP_ID,PERP_AGE_GROUP,PERP_SEX,PERP_RACE,VICTIM_ID,VICTIM_AGE_GROUP,VICTIM_SEX,VICTIM_RACE,STAT_MURDER_FLG
0,9953245,01/01/2006,02:00:00,BRONX,OUTSIDE,48.0,0.0,STREET,NONE,1013404.0,...,-73.894607,9953245-9037,18-24,MALE,BLACK,9953245-8156,<18,MALE,BLACK,N
1,9953246,01/01/2006,05:51:00,BRONX,OUTSIDE,44.0,0.0,STREET,NONE,1007446.0,...,-73.916173,9953246-9039,25-44,MALE,WHITE HISPANIC,9953246-8158,18-24,MALE,WHITE HISPANIC,N
2,9953247,01/01/2006,03:30:00,BROOKLYN,OUTSIDE,67.0,0.0,STREET,NaN,999316.0,...,-73.945705,9953247-9041,UNKNOWN,UNKNOWN,UNKNOWN,9953247-8160,18-24,MALE,BLACK,N
3,9953247,01/01/2006,03:30:00,BROOKLYN,OUTSIDE,67.0,0.0,STREET,NaN,999316.0,...,-73.945705,9953247-9041,UNKNOWN,UNKNOWN,UNKNOWN,9953247-8161,18-24,MALE,BLACK,N
4,9953247,01/01/2006,03:30:00,BROOKLYN,OUTSIDE,67.0,0.0,STREET,NaN,999316.0,...,-73.945705,9953247-9041,UNKNOWN,UNKNOWN,UNKNOWN,9953247-8162,18-24,MALE,BLACK,N
5,9953248,01/01/2006,19:00:00,QUEENS,OUTSIDE,106.0,0.0,STREET,NONE,1028607.0,...,-73.840072,9953248-9042,18-24,MALE,BLACK,9953248-8163,18-24,MALE,BLACK,N
6,9953249,01/02/2006,00:49:00,BROOKLYN,OUTSIDE,90.0,2.0,HOUSING,MULTI DWELL - PUBLIC HOUS,1000067.0,...,-73.942948,9953249-9043,UNKNOWN,UNKNOWN,UNKNOWN,9953249-8164,25-44,MALE,BLACK,Y
7,9953250,01/01/2006,02:34:00,QUEENS,OUTSIDE,114.0,0.0,COMMERCIAL,BAR/NIGHT CLUB,1002576.0,...,-73.933833,9953250-9044,25-44,MALE,BLACK,9953250-8165,25-44,MALE,BLACK,Y
8,9953250,01/01/2006,02:34:00,QUEENS,OUTSIDE,114.0,0.0,COMMERCIAL,BAR/NIGHT CLUB,1002576.0,...,-73.933833,9953250-9044,25-44,MALE,BLACK,9953250-8166,25-44,MALE,BLACK,Y
9,9953250,01/01/2006,02:34:00,QUEENS,OUTSIDE,114.0,0.0,COMMERCIAL,BAR/NIGHT CLUB,1002576.0,...,-73.933833,9953250-9196,18-24,MALE,BLACK,9953250-8165,25-44,MALE,BLACK,Y


In [69]:
shootings_df.head(10)

,INCIDENT_KEY,OCCUR_DATE,OCCUR_TIME,BORO,LOC_OF_OCCUR_DESC,PRECINCT,JURISDICTION_CODE,LOC_CLASSFCTN_DESC,LOCATION_DESC,X_COORD_CD,Y_COORD_CD,Latitude,Longitude
0,297623042,12/06/2024,13:06:00,MANHATTAN,OUTSIDE,1,0.0,STREET,NaN,983437,201643,40.720149,-74.002931
1,73794477,07/21/2010,01:32:00,MANHATTAN,OUTSIDE,1,0.0,STREET,BAR/NIGHT CLUB,982761,204594,40.728241,-74.005370
2,242784290,03/28/2022,22:18:00,MANHATTAN,OUTSIDE,5,0.0,STREET,MULTI DWELL - APT BUILD,984732,200728,40.717629,-73.998260
3,24449553,09/23/2006,00:02:00,MANHATTAN,OUTSIDE,5,0.0,STREET,NaN,986929,202081,40.721343,-73.990335
4,64318199,08/02/2009,01:44:00,MANHATTAN,OUTSIDE,5,2.0,HOUSING,MULTI DWELL - PUBLIC HOUS,986577,200423,40.716801,-73.991606
5,11118235,04/26/2006,21:05:00,MANHATTAN,OUTSIDE,6,0.0,STREET,NaN,982301,204919,40.729142,-74.007033
6,235033374,10/15/2021,20:30:00,MANHATTAN,OUTSIDE,7,0.0,STREET,NaN,987969,200865,40.718004,-73.986584
7,52144006,10/10/2008,01:38:00,MANHATTAN,OUTSIDE,7,2.0,HOUSING,MULTI DWELL - PUBLIC HOUS,989248,198472,40.711446,-73.981970
8,37401695,01/01/2008,05:12:00,MANHATTAN,OUTSIDE,7,2.0,HOUSING,MULTI DWELL - PUBLIC HOUS,989138,199179,40.713385,-73.982369
9,16631507,05/31/2006,21:16:00,MANHATTAN,OUTSIDE,7,2.0,HOUSING,MULTI DWELL - APT BUILD,989248,198472,40.711446,-73.981970


In [70]:
offenders_df.head(10)

,INCIDENT_KEY,PERP_ID,PERP_AGE_GROUP,PERP_SEX,PERP_RACE
0,9953248,9953248-9042,18-24,MALE,BLACK
1,10137418,10137418-9127,25-44,MALE,BLACK
2,10137428,10137428-9147,UNKNOWN,UNKNOWN,UNKNOWN
3,10187494,10187494-9181,UNKNOWN,MALE,BLACK
4,10187494,10187494-9195,UNKNOWN,MALE,BLACK
5,10187496,10187496-9184,18-24,MALE,WHITE HISPANIC
6,10187501,10187501-9189,<18,MALE,BLACK
7,10187502,10187502-9190,25-44,MALE,BLACK
8,10187517,10187517-9212,25-44,MALE,BLACK
9,10240242,10240242-9227,18-24,MALE,WHITE HISPANIC


In [71]:
victims_df.head(10)

,INCIDENT_KEY,VICTIM_ID,VICTIM_AGE_GROUP,VICTIM_SEX,VICTIM_RACE,STAT_MURDER_FLG
0,299453927,299453927-1,25-44,MALE,BLACK,N
1,212336415,212336415-1,18-24,MALE,BLACK,N
2,280768895,280768895-1,25-44,MALE,BLACK,N
3,141824677,141824677-27687,25-44,MALE,BLACK HISPANIC,N
4,270079203,270079203-1,45-64,MALE,BLACK,Y
5,138882700,138882700-26963,25-44,MALE,BLACK,N
6,138882700,138882700-26964,45-64,MALE,BLACK,N
7,289881880,289881880-1,25-44,MALE,WHITE HISPANIC,N
8,305930898,305930898-1,25-44,MALE,BLACK,N
9,138899428,138899428-26973,18-24,MALE,BLACK,N


In [72]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34148 entries, 0 to 34147
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   INCIDENT_KEY        34148 non-null  int64  
 1   OCCUR_DATE          34145 non-null  object 
 2   OCCUR_TIME          34145 non-null  object 
 3   BORO                34145 non-null  object 
 4   LOC_OF_OCCUR_DESC   34145 non-null  object 
 5   PRECINCT            34145 non-null  float64
 6   JURISDICTION_CODE   34143 non-null  float64
 7   LOC_CLASSFCTN_DESC  34105 non-null  object 
 8   LOCATION_DESC       13928 non-null  object 
 9   X_COORD_CD          34145 non-null  float64
 10  Y_COORD_CD          34145 non-null  float64
 11  Latitude            33966 non-null  float64
 12  Longitude           33966 non-null  float64
 13  PERP_ID             23096 non-null  object 
 14  PERP_AGE_GROUP      23061 non-null  object 
 15  PERP_SEX            23096 non-null  object 
 16  PERP

In [73]:
combined_df.describe()

,INCIDENT_KEY,PRECINCT,JURISDICTION_CODE,X_COORD_CD,Y_COORD_CD,Latitude,Longitude
count,3.414800e+04,34145.000000,34143.000000,3.414500e+04,34145.000000,33966.000000,33966.000000
mean,1.383745e+08,65.372763,0.312831,1.009558e+06,208677.347987,40.667336,-73.778148
std,8.750975e+07,27.631678,0.724106,1.842505e+04,31915.912184,1.712999,3.104352
min,9.953245e+06,1.000000,0.000000,9.148030e+05,125922.000000,0.000000,-74.249755
25%,6.652822e+07,44.000000,0.000000,1.000094e+06,182881.000000,40.668341,-73.942753
50%,1.373663e+08,67.000000,0.000000,1.007847e+06,195571.000000,40.702886,-73.914719
75%,2.176082e+08,81.000000,0.000000,1.016871e+06,239961.000000,40.825169,-73.882032
max,3.182339e+08,123.000000,2.000000,1.066815e+06,271127.000000,40.910810,0.000000


In [74]:
print('Duplicate Values in Combined Dataset: ', combined_df.duplicated().sum())

Duplicate Values in Combined Dataset:  0


In [78]:
print('Null Values in Combined Dataset: ')
combined_df.isna().sum()

Null Values in Combined Dataset: 


INCIDENT_KEY              0
OCCUR_DATE                3
OCCUR_TIME                3
BORO                      3
LOC_OF_OCCUR_DESC         3
PRECINCT                  3
JURISDICTION_CODE         5
LOC_CLASSFCTN_DESC       43
LOCATION_DESC         20220
X_COORD_CD                3
Y_COORD_CD                3
Latitude                182
Longitude               182
PERP_ID               11052
PERP_AGE_GROUP        11087
PERP_SEX              11052
PERP_RACE             11052
VICTIM_ID                 1
VICTIM_AGE_GROUP          1
VICTIM_SEX                1
VICTIM_RACE               1
STAT_MURDER_FLG           1
dtype: int64

In [75]:
print('Shape of Combined Dataset: ',combined_df.shape)
print('Shape of Shootings Dataset: ',shootings_df.shape)
print('Shape of Offenders Dataset: ',offenders_df.shape)
print('Shape of Victims Dataset: ',victims_df.shape)

Shape of Combined Dataset:  (34148, 22)
Shape of Shootings Dataset:  (23988, 13)
Shape of Offenders Dataset:  (18750, 5)
Shape of Victims Dataset:  (28748, 6)


## Observations

- There are atleast **11052** records in Combined data, details of the offenders are missing, either because the offenders were never recorded or the incident details don't exist in the offenders dataset.

- There are **20220** location descriptions missing.

- `Latitude` and `Longitude` have more missing values than `X_COORD_CD` and `Y_COORD_CD`, when both are supposed to have information about the location incident, it could be a data entry issue, could it be that the `Latitude` and `Longitude` were introduced to recording practice later.

- From the initial head view, both victims(`VICTIM_RACE`) and offenders(`PERP_RACE`) are highly saturated with `BLACK` and `HISPANIC` values, which is consistent with the high percentage(~48%) of African American and Hispanic in NYC.

- From the initial look, the age groups `18-24` and `25-44` are dominant in the dataset, which aligns with these being the largest adult demographic groups in NYC's population

- From the initial head view, both victims and offenders appear to be largely male across the datasets.

## Questions

- What time of the day, does shootings mostly occur?

- What part of NYC have the highest number of shooting incidents?

- Have shooting incidents decrease over time, both city-wide or by borough?

- Does the location of the incident relate to whether the shooting is fatal or not?